android/modules/ui/src/java/ti/modules/titanium/ui/widget/webview/TiWebViewBinding.java

```diff
@@ -217,32 +253,69 @@ public void call(Object data)
 		}
 	}
 
+	private static class Event
+	{
+		public String name;
+		public KrollDict properties;
+	}
+
 	@SuppressWarnings("unused")
 	private class AppBinding
 	{
 		private KrollModule module;
-		private HashMap<String, Integer> appListeners = new HashMap<String, Integer>();
+		private HashMap<String, Integer> appListeners = new HashMap<>();
+		private ArrayDeque<Event> eventQueue = new ArrayDeque<>();
 		private int counter = 0;
 		private String code = null;
+
 		public AppBinding()
 		{
 			module = TiApplication.getInstance().getModuleByName("App");
 		}
 
 		@JavascriptInterface
-		public void fireEvent(String event, String json)
+		public void fireEvent(String eventName, String json)
 		{
 			try {
-				KrollDict dict = new KrollDict();
+				// Parse event properties from given JSON and store to event object.
+				Event event = new Event();
+				event.name = eventName;
+				event.properties = new KrollDict();
 				if (json != null && !json.equals("undefined")) {
-					dict = new KrollDict(new JSONObject(json));
+					event.properties = new KrollDict(new JSONObject(json));
+				}
+
+				// Queue event and fire it, if ready.
+				synchronized (this)
```

Comment:
I don't think the `synchronized` is necessary?

Reply:
This method can be called by the WebView's JavaScript thread (via `@JavaScriptInterface` APIs) or the main UI thread (via `resumeFiringEvents()` calls). So, the synchronized is needed. And we shouldn't invoke a listener within a synchronized block to avoid potential deadlock.